### DISCUSSION TITLE
Running a tag filter on all tags, then running another on all tags again.

### DISCUSSION THREAD

**Discussion by SychO9** [s9e/TextFormatter#174] [Support]
Hello 👋🏼,

First of all, incredible library! I read through the docs and couldn't find what I'm looking for so not sure if it's possible or not, but wanted to ask away.

Basically, I have a tag configured with an `ID` attribute and a regular expression to extract it from, and I'm running a filter on this tag where the `ID` is used to identify a database model and validate the tag depending on its existence. What I'm trying to do is run another filter first, but run it on all tag instances before running the other filter on all of them again so that I can collect the caught IDs first before doing anything. I might've missed something very simple here but oh well.

Thank you very much.

**Comment by JoshyPHP** [comment]
Glad you like it. I'm trying to understand how your case should work. Let's say that your tag is named `X`, IDs are stored as `id` attributes and we use BBCodes for simplicity. Tags with `id === 1` are valid, others are not. We get this code:
```php
$configurator = new s9e\TextFormatter\Configurator;
$configurator->BBCodes->addCustom('[x id={INT}]', '');

$configurator->tags['X']->attributes['id']->filterChain->append('validateId');

function validateId(int $id)
{
	return ($id === 1);
}

extract($configurator->finalize());

$text = '[x=1] [x=2]';
$xml  = $parser->parse($text);

die("$xml\n");
```
```xml
<r><X id="1">[x=1]</X> [x=2]</r>
```

Do you want to run a function with all IDs (so an array such as `[1, 2]`) before `validateId()` is called for the first time? Is it because you want to avoid running N queries and would rather prefetch everything with a single query? Or are you trying to do something different?

**Reply by SychO9** [reply]
> Do you want to run a function with all IDs (so an array such as `[1, 2]`) before `validateId()` is called for the first time? Is it because you want to avoid running N queries and would rather prefetch everything with a single query?

Yes! precisely that! with user input it can easily get out of hand, and I'd like to avoid that, so any way of being able to get all those IDs first to prefetch everything would be perfect, though I'm still clueless if that's possible or not.

**Reply by JoshyPHP** [reply]
Bad news then, because that specific usage isn't supported right now. You can't access tags outside of the filtering process, and filters only have access to the current tag and its ancestors. So no siblings or descendants.

I have considered various solutions to handle cases like yours but none that seemed like a definite winner, that's why I haven't committed to any. At the moment, all I can offer are possible workarounds.


### Workaround #1: roll-your-own prefetcher

You can extract the IDs manually, with some sort of `preg_match_all()` that you run before parsing. You'd store their status in an object that is passed to a filter. In the following example, I manually create an `ArrayObject` instance and register it as `cache`. That means it becomes available in the callback's short syntax as `$cache`.

 * https://s9etextformatter.readthedocs.io/Filters/Callback_signature/

```php
$configurator = new s9e\TextFormatter\Configurator;
$configurator->BBCodes->addCustom('[x id={INT}]', '');

$configurator->tags['X']->attributes['id']->filterChain->append('validateId($attrValue, $cache)');

function validateId(int $id, $cache)
{
	return isset($cache[$id]);
}

extract($configurator->finalize());

// Create the cache instance and register it into the parser
$cache    = new ArrayObject;
$cache[1] = [];
$parser->registeredVars['cache'] = $cache;

$text = '[x=1] [x=2]';
$xml  = $parser->parse($text);

die("$xml\n");
```

The downside is that you have to extract those IDs manually. You get no help from the library


### Workaround #2: roll-your-own post-processor

Alternatively, you can accept all instances of `X` and add a flag (an attribute) that indicates whether it's valid. That flag is left empty by default, and you post-process the XML output to set it as appropriate.

```php
use s9e\TextFormatter\Utils;

$configurator = new s9e\TextFormatter\Configurator;

// Use the #false built-in filter to systematically discard user input
$configurator->BBCodes->addCustom(
	'[x id={INT} valid={TEXT?;postFilter=#false}]',
	'<xsl:choose>
		<xsl:when test="@valid">Valid id={INT}</xsl:when>
		<xsl:otherwise>Invalid id={INT}</xsl:otherwise>
	</xsl:choose>'
);

extract($configurator->finalize());

$text = '[x=1] [x=2]';
$xml  = $parser->parse($text);

echo "Before:\n$xml\n\n";

// Retrieve the IDs from the document
$ids = Utils::getAttributeValues($xml, 'X', 'id');

// Fetch from the database, but here we'll hardcode it
$cache = [1 => true];

// Iterate over all X tags, set the "valid" flag where appropriate
$xml = Utils::replaceAttributes(
	$xml,
	'X',
	function ($attributes) use ($cache): array
	{
		if (isset($attributes['id'], $cache[$attributes['id']]))
		{
			$attributes['valid'] = '';
		}

		return $attributes;
	}
);

$html = $renderer->render($xml);
echo "After:\n$xml\n\nHTML:\n$html\n";
```
```
Before:
<r><X id="1">[x=1]</X> <X id="2">[x=2]</X></r>

After:
<r><X id="1" valid="">[x=1]</X> <X id="2">[x=2]</X></r>

HTML:
Valid id=1 Invalid id=2
```

The downside is that you end up with unwanted tags in the XML because there is no built-in functionality to remove tags that don't have the `valid` attribute, but you could remove them using normal DOM manipulation. In that case, you'd probably be better off just collecting the IDs with `Utils::getAttributeValues()` then use DOM to iterate over all `X` tags to remove them rather than flag them *then* remove then.

Oh, and also if you go with post-processing it means that you don't know whether a tag is actually valid *during* parsing. That may be relevant if that tag has descendants.

**Reply by SychO9** [reply]
Thank you for the detailed answer, much appreciated.

I had implemented the second work around where I left the attributes that rely on the `ID` attribute empty and had them filled in the rendering process where models were prefetched. But as you've mentioned, that ends up with invalid tags in the XML, so before starting to mess with that I wanted to be 100% certain there wasn't a better way through the library.

Though the first workaround seems like a better option, since I'm already using a regular expression to catch the `ID` attribute, I could run it manually first, prefetch everything I need and read from that.

I'll play around with both solutions and see what fits best :)

Again, thank you very much, I got the answer I needed, and thank you for an awesome library!
